gems/smithy/lib/smithy/templates/client/protocol_plugin.erb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+# This is generated code!
+
+module <%= module_name %>
+  module Plugins
+     # Protocol plugin
+     class Protocol < Smithy::Client::Plugin
+       option(
+         :protocol,
+         doc_default: 'Smithy::Client::Protocols::<%= protocol_name %>',
+         doc_type: 'Smithy::Client::Protocols::ProtocolBase',
+         docstring: 'TODO'
+         ) do |_cfg|
+           Smithy::Client::Protocols::<%= protocol_name %>.new
```

Comment:
I understand why we're using a codegen plugin to determine the default protocol here and this is pretty reasonable.

I was trying to think through alternatives to this - one option would be a `ProtocolResolver` that takes the schema (and its associated, ordered protocol traits) and then loads all of the descendants of ProtocolBase and finds the first matching one.  Sort of like engines, this would let users dynamically pick protocols based on 1) the preference of the service and 2) what they have available at runtime.

The downside of that is that the default protocol used would change based on what the user has available... and means we couldn't give it a good doc default either.

Reply:
I don't exactly recall how flexible protocol swapping should happen. It didn't occur to me until now that only protocol traits defined on the service shape could be allowed as protocol alternatives. I was under the impression that customers could bring ANY protocol of their choice - outside of the service shape model...

Let's say that service shape traits only lists one protocol trait. Generated client with using that protocol as default. Customer wrote their own protocol (follows to our protocol implementation guidance) and decides to pass it in the client config  - is this allowed?

Reply:
I think so. They could modify an existing protocol to have separate behavior. I don't think it's useful other than cases where we make a backwards incompatible change or something.